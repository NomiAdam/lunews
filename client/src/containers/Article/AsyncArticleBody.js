import React from 'react';
import PropTypes from 'prop-types';
import { prop, ifElse } from 'ramda';
import { alwaysNull } from 'ramda-extension';
import { Grid, GridCol } from 'bachelor-ui';
import SuspenseLoader from '../../components/SuspenseLoader';

const bodyStyle = {
    padding: '20px 0',
};

const AsyncArticleBody = ifElse(
    prop('body'),
    ({ body }) => (
        <Grid>
            <GridCol colXS={ 12 }>
                <div style={ bodyStyle } dangerouslySetInnerHTML={ { __html: body } } />
            </GridCol>
        </Grid>
    ),
    alwaysNull,
);

AsyncArticleBody.propTypes = {
    body: PropTypes.string,
};

const AsyncTreeBody = ({ body }) => (
    <SuspenseLoader>
        <AsyncArticleBody body={ body } />
    </SuspenseLoader>
);

AsyncTreeBody.propTypes = {
    body: PropTypes.string,
};

export default AsyncTreeBody;
