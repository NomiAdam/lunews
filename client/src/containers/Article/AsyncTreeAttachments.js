import React from 'react';
import PropTypes from 'prop-types';
import {
    Grid, GridCol, Paragraph, Icon, Container, Media, flexConstants, lightTheme,
} from 'bachelor-ui';
import {
    map, filter, propEq, anyPass, o, propOr, path, not, prop, ifElse,
} from 'ramda';
import { alwaysNull } from 'ramda-extension';
import { saveAs } from 'file-saver';
import SuspenseLoader from '../../components/SuspenseLoader';

const propEqualImageJpeg = propEq('contentType', 'image/jpeg');
const propEqualImageJpg = propEq('contentType', 'image/jpg');
const propEqualImagePng = propEq('contentType', 'image/png');
const isImage = anyPass([propEqualImageJpeg, propEqualImageJpg, propEqualImagePng]);

const filterOutImages = filter(isImage);
const filterOutOtherFiles = filter(o(not, isImage));

const fileNameOrDefaultP = propOr('filename', 'filename');
const contentTypeP = propOr('', 'contentType');
const contentDataP = path(['content', 'data']);
const renderImageGridCol = map(
    attachment => (
        <GridCol
            key={ fileNameOrDefaultP(attachment) }
            colXS={ 12 }
            colMD={ 4 }
            horizontalAlign={ flexConstants.POSITION_LEFT }
            verticalAlign={ flexConstants.POSITION_CENTER }
        >
            <Container
                flexDirection={ flexConstants.DIRECTION_COLUMN }
                backgroundColor={ lightTheme.transparent }
                overflow="hidden"
            >
                <Paragraph>{fileNameOrDefaultP(attachment)}</Paragraph>
                <Media
                    onClick={ (e) => {
                        e.stopPropagation();
                        const byteArray = new Uint8Array(contentDataP(attachment));
                        const blob = new Blob([byteArray], { type: contentTypeP(attachment) });
                        saveAs(blob, fileNameOrDefaultP(attachment));
                    } }
                    alt={ fileNameOrDefaultP(attachment) }
                    image={ `data:${ contentTypeP(attachment) };base64,${ btoa(new Uint8Array(contentDataP(attachment)).reduce((data, byte) => data + String.fromCharCode(byte), '')) }` }
                />
            </Container>
        </GridCol>
    ),
);

const renderImageAttachments = o(renderImageGridCol, filterOutImages);

const renderOtherGridCol = map(
    attachment => (
        <GridCol
            key={ fileNameOrDefaultP(attachment) }
            colXS={ 12 }
            colMD={ 4 }
            horizontalAlign={ flexConstants.POSITION_LEFT }
            verticalAlign={ flexConstants.POSITION_CENTER }
        >
            <Container
                padding="0px 10px"
                flexDirection={ flexConstants.DIRECTION_ROW }
                justifyContent={ flexConstants.POSITION_SPACER }
                backgroundColor={ lightTheme.transparent }
            >
                <Paragraph>
                    {fileNameOrDefaultP(attachment)}
                </Paragraph>
                <Icon
                    name="download"
                    secondary
                    clickable
                    onClick={ (e) => {
                        e.stopPropagation();
                        const byteArray = new Uint8Array(contentDataP(attachment));
                        const blob = new Blob([byteArray], { type: contentTypeP(attachment) });
                        saveAs(blob, fileNameOrDefaultP(attachment));
                    } }
                />
            </Container>
        </GridCol>
    ),
);
const renderOtherAttachments = o(renderOtherGridCol, filterOutOtherFiles);

const AsyncAttachments = ifElse(
    prop('attachments'),
    ({ attachments }) => (
        <Grid gridGap="10px">
            {
                renderImageAttachments(attachments)
            }
            {
                renderOtherAttachments(attachments)
            }
        </Grid>
    ),
    alwaysNull,
);

AsyncAttachments.propTypes = {
    AsyncAttachments: PropTypes.array,
};

const AsyncTreeAttachments = ({ attachments }) => (
    <SuspenseLoader>
        <AsyncAttachments attachments={ attachments } />
    </SuspenseLoader>
);

AsyncTreeAttachments.propTypes = {
    attachments: PropTypes.array,
};

export default AsyncTreeAttachments;
