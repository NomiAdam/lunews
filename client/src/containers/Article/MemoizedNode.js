import React, { Fragment, memo } from 'react';
import PropTypes from 'prop-types';
import {
    basicTheme, GridCol, Heading, Icon, Paragraph,
} from 'bachelor-ui';
import { Link } from 'react-router-dom';
import flexConstants from 'bachelor-ui/dist/constants/flex';
import formatDate from '../../utils/formatDate';
import { setReply } from '../../actions/uiActions';

const linkStyle = {
    outline: 'none',
    textDecoration: 'none',
};

const MemoizedNode = ({
    subject, date, sender, isFocused, attachmentsSize, setReplyTrue, hasAttachments, downloadFiles,
    isStarred, removeRestFavourite, setRestFavourite, isMobile, iconGridSize,
}) => {
    const IconGridCol = memo(({ children }) => (
        <GridCol
            colXS={ iconGridSize }
            verticalAlign={ flexConstants.POSITION_CENTER }
            horizontalAlign={ flexConstants.POSITION_CENTER }
        >
            { children }
        </GridCol>
    ));
    return (
        <Fragment>
            <GridCol colXS={ 12 }>
                <Heading secondary label={ subject } />
            </GridCol>
            <GridCol colXS={ 12 }>
                <Paragraph
                    color={ basicTheme.grey }
                    size="0.8em"
                >
                    { formatDate(date) }
                </Paragraph>
            </GridCol>
            <GridCol colXS={ 12 }>
                <Paragraph margin="20px 0" size="1.8m" secondary>{ sender }</Paragraph>
            </GridCol>
            {isFocused && (
                <Fragment>
                    <IconGridCol>
                        <Icon
                            name="attach"
                            clickable
                            fontSize="1.5em"
                            secondary
                            onClick={ setReply }
                        />
                        <Paragraph bold size="1.5em">{ attachmentsSize }</Paragraph>
                    </IconGridCol>
                    <IconGridCol>
                        {
                            isStarred ? (
                                <Icon
                                    name="starFill"
                                    secondary
                                    clickable
                                    fontSize="1.5em"
                                    onClick={ removeRestFavourite }
                                />
                            ) : (
                                <Icon
                                    name="star"
                                    clickable
                                    fontSize="1.5em"
                                    onClick={ setRestFavourite }
                                />
                            )
                        }
                    </IconGridCol>
                    <IconGridCol>
                        {
                            isMobile ? (
                                <Link to="/post" style={ linkStyle }>
                                    <Icon name="reply" fontSize="2em" clickable secondary />
                                </Link>
                            ) : (
                                <Icon
                                    name="reply"
                                    fontSize="1.5em"
                                    secondary
                                    clickable
                                    onClick={ setReplyTrue }
                                />
                            )
                        }
                    </IconGridCol>
                    {
                        hasAttachments && (
                            <IconGridCol>
                                <Icon
                                    name="download"
                                    color="red"
                                    clickable
                                    onClick={ downloadFiles }
                                />
                            </IconGridCol>
                        )
                    }
                </Fragment>
            )}
        </Fragment>
    );
};

MemoizedNode.propTypes = {
    subject: PropTypes.string,
    date: PropTypes.string,
    sender: PropTypes.string,
    isFocused: PropTypes.bool,
    attachmentsSize: PropTypes.number,
    setReplyTrue: PropTypes.func,
    hasAttachments: PropTypes.bool,
    downloadFiles: PropTypes.func,
    isStarred: PropTypes.bool,
    removeRestFavourite: PropTypes.func,
    setRestFavourite: PropTypes.func,
    isMobile: PropTypes.bool,
    iconGridSize: PropTypes.number,
};

export default MemoizedNode;
