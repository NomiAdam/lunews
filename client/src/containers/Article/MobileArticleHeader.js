import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';
import Header from '../../components/Header';
import BackArrow from '../../components/BackArrow';

const MobileArticleHeader = ({ history: { push } }) => {
    const redirectBack = useRef(() => push('/list'));
    return (
        <Header
            before
            actionChildren={ (
                <BackArrow
                    handleBack={ redirectBack.current }
                    label="Article"
                />
            ) }
        />
    );
};

MobileArticleHeader.propTypes = {
    history: PropTypes.any,
};

export default withRouter(MobileArticleHeader);
