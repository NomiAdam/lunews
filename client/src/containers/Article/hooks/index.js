import useFetchArticleTree from './useFetchArticleTree';
import useGetRestArticle from './useGetRestArticle';
import useIsFocused from './useIsFocused';
import useFetchAttachments from './useFetchAttachments';
import useRestFavourite from './useRestFavourite';
import useFetchBody from './useFetchBody';

export {
    useFetchAttachments,
    useRestFavourite,
    useFetchArticleTree,
    useIsFocused,
    useGetRestArticle,
    useFetchBody,
};
