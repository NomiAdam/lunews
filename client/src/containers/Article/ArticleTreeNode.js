import React, { useContext, useRef, useCallback } from 'react';
import PropTypes from 'prop-types';
import { Grid, GridCol } from 'bachelor-ui';
import { saveAs } from 'file-saver';
import { forEach, o } from 'ramda';
import { defaultToEmptyArray, isNotEmpty } from 'ramda-extension';
import { GroupContext, UiContext } from '../../context';
import {
    useIsFocused, useRestFavourite, useFetchAttachments, useFetchBody,
} from './hooks';
import { setReply } from '../../actions/uiActions';
import AsyncTreeAttachments from './AsyncTreeAttachments';
import AsyncArticleBody from './AsyncArticleBody';
import MemoizedNode from './MemoizedNode';

const downloadAttachment = (attachment) => {
    const byteArray = new Uint8Array(attachment.content.data);
    const blob = new Blob([byteArray], { type: attachment.contentType });
    saveAs(blob, attachment.filename);
};
const downloadAllFiles = o(forEach(downloadAttachment), defaultToEmptyArray);

const ArticleTreeNode = ({
    sender, date, globalID, isMobile, subject,
}) => {
    const { dispatch } = useContext(UiContext);
    const { group: { isStarred } } = useContext(GroupContext);
    const [isFocused] = useIsFocused(globalID);
    const [setRestFavourite, removeRestFavourite] = useRestFavourite(globalID);
    const { attachments, attachmentsSize } = useFetchAttachments(globalID);
    const [body] = useFetchBody(globalID);
    const hasAttachments = isNotEmpty(attachments);
    const iconGridSize = hasAttachments ? 3 : 4;

    const setReplyTrue = useRef(() => dispatch(setReply(true)));
    const downloadFiles = useCallback(() => downloadAllFiles(attachments), [JSON.stringify(attachments)]);

    return (
        <Grid>
            <MemoizedNode
                attachmentsSize={ attachmentsSize }
                date={ date }
                downloadFiles={ downloadFiles }
                hasAttachments={ hasAttachments }
                iconGridSize={ iconGridSize }
                isFocused={ isFocused }
                isMobile={ isMobile }
                isStarred={ isStarred }
                removeRestFavourite={ removeRestFavourite }
                sender={ sender }
                setReplyTrue={ setReplyTrue.current }
                setRestFavourite={ setRestFavourite }
                subject={ subject }
            />
            <GridCol colXS={ 12 }>
                { isFocused && <AsyncArticleBody body={ body } /> }
            </GridCol>
            <GridCol colXS={ 12 }>
                { isFocused && <AsyncTreeAttachments attachments={ attachments } /> }
            </GridCol>
        </Grid>
    );
};

ArticleTreeNode.propTypes = {
    sender: PropTypes.string,
    date: PropTypes.string,
    globalID: PropTypes.string,
    isMobile: PropTypes.bool,
    subject: PropTypes.string,
};

export default ArticleTreeNode;
