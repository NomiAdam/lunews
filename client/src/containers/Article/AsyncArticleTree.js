/* eslint-disable no-nested-ternary */
import React, { useContext, useRef } from 'react';
import PropTypes from 'prop-types';
import { o } from 'ramda';
import {
    TreeList, Icon, Container, flexConstants, Paper, Heading,
} from 'bachelor-ui';
import ArticleTreeNode from './ArticleTreeNode';
import { useGetRestArticle, useFetchArticleTree } from './hooks';
import { setFocusedMessage, setActiveMessage } from '../../actions/groupActions';
import { GroupContext } from '../../context';
import FakeLoader from '../../components/FakeLoader';

const AsyncArticleTree = ({ isMobile }) => {
    const { dispatch: GroupDispatch } = useContext(GroupContext);
    const clearMessage = useRef(() => GroupDispatch(setActiveMessage(undefined)));
    const setMessage = useRef(o(GroupDispatch, setFocusedMessage));
    const [articleTree, isLoading] = useFetchArticleTree();
    useGetRestArticle();
    return articleTree ? (
        <Container
            width="100%"
            padding="5px 10px"
            flexDirection={ flexConstants.DIRECTION_COLUMN }
            alignItems={ flexConstants.POSITION_LEFT }
        >
            {
                !isMobile && (
                    <Icon
                        name="close"
                        secondary
                        clickable
                        onClick={ clearMessage.current }
                    />
                )
            }
            <TreeList
                clickable
                treeData={ articleTree.tree }
                dataProp="messageInfo"
                handleClick={ setMessage.current }
                node={ ({
                    sender, date, globalID, subject,
                }) => (
                    <ArticleTreeNode
                        globalID={ globalID }
                        subject={ subject }
                        date={ date }
                        isMobile={ isMobile }
                        sender={ sender }
                    />
                )
                }
            />
        </Container>
    ) : isLoading ? <FakeLoader /> : (
        <Container
            width="100%"
            padding="5px 10px"
            flexDirection={ flexConstants.DIRECTION_COLUMN }
            alignItems={ flexConstants.POSITION_LEFT }
        >
            {
                !isMobile && (
                    <Icon
                        name="close"
                        secondary
                        clickable
                        onClick={ clearMessage.current }
                    />
                )
            }
            <Paper width="100%">
                <Heading label="Article not found." />
            </Paper>
        </Container>
    );
};

AsyncArticleTree.propTypes = {
    isMobile: PropTypes.bool,
};

export default AsyncArticleTree;
