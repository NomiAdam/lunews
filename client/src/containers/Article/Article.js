import React from 'react';
import { Container, flexConstants } from 'bachelor-ui';
import AsyncArticleTree from './AsyncArticleTree';
import SuspenseLoader from '../../components/SuspenseLoader';

const Article = ({ isMobile }) => (
    <Container
        height={ isMobile ? '90vh' : '100vh' }
        flexDirection={ flexConstants.DIRECTION_COLUMN }
        alignItems={ flexConstants.POSITION_RIGHT }
        justifyContent={ flexConstants.POSITION_LEFT }
    >
        <SuspenseLoader>
            <AsyncArticleTree isMobile={ isMobile } />
        </SuspenseLoader>
    </Container>
);

export default Article;
