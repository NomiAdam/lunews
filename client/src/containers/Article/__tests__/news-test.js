/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import { Article } from '../Article';

const mockFn = () => new Promise(resolve => resolve);

describe('Article', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });
    it('should be defined', () => {
        expect(Article)
            .toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<Article fetchThread={ mockFn } />))
            .toMatchSnapshot();
    });
    it('should have default state defined', () => {
        expect(shallow(<Article
            match={ {
                params: {
                    globalID: 'mock',
                },
            } }
            fetchThread={ mockFn }
        />)
            .state())
            .toEqual({
                article: {
                    messageID: '',
                    lines: [],
                },
                replies: [],
                reference: undefined,
                globalID: 'mock',
                openedReply: false,
            });
    });
});
