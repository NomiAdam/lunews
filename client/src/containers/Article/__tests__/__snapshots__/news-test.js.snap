// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article should match the snapshot 1`] = `
<div>
  <OverlayHeading
    handleIconClick={[Function]}
    heading="Article"
    isOpen={false}
    verticalPosition="center"
  />
  <NewsTree
    article={
      Object {
        "lines": Array [],
        "messageID": "",
      }
    }
    dispatchReply={[Function]}
    replies={Array []}
  />
  <Spacer
    height="63px"
    width="100%"
  />
  <OverlayForm
    handleClose={[Function]}
    handlePost={[Function]}
    isOpen={false}
  />
</div>
`;
